acados/ocp_nlp/ocp_nlp_sqp.c

```diff
@@ -653,10 +733,573 @@ static void ocp_nlp_sqp_reset_timers(ocp_nlp_sqp_memory *mem)
+                {
+                    debug_output(opts->nlp_opts, "f-type step: Armijo condition NOT satisfied\n", 1); //debugging output
+                }
+
+            }
+            else if (is_funnel_sufficient_decrease_satisfied(mem, opts, trial_infeasibility))
+            {
+                debug_output(opts->nlp_opts, "Switching condition is NOT satisfied!\n", 1); //debugging output
+                debug_output(opts->nlp_opts, "h-type step: funnel suff. decrease satisfied!\n", 1); //debugging output
+                accept_step = true;
+                mem->funnel_iter_type = 'h';
+                decrease_funnel(mem, opts, trial_infeasibility, current_infeasibility);
+            }
+            else
+            {
+                debug_output(opts->nlp_opts, "Switching condition is NOT satisfied!\n", 1); //debugging output
+                debug_output(opts->nlp_opts, "Entered penalty check!\n", 1); //debugging output
+                //TODO move to function and test more
+                if (trial_merit <= current_merit + opts->linesearch_eta * alpha * pred_merit)
+                {
+                    debug_output(opts->nlp_opts, "Penalty Function accepted\n", 1);
+                    accept_step = true;
+                    mem->funnel_iter_type = 'b';
+                    mem->funnel_penalty_mode = true;
+                }
+            }
+        }
+        else
+        {
+            debug_output(opts->nlp_opts, "Penalty mode active\n", 1);
+            if (trial_merit <= current_merit + opts->linesearch_eta * alpha * pred_merit)
+            {
+                debug_output(opts->nlp_opts, "p-type step: accepted iterate\n", 1);
+                accept_step = true;
+                mem->funnel_iter_type = 'p';
+
+                if (is_funnel_sufficient_decrease_satisfied(mem, opts, trial_infeasibility))
+                {
+                    decrease_funnel(mem, opts, trial_infeasibility, current_infeasibility);
+                    mem->funnel_penalty_mode = false;
+                }
+            }
+        }
+    }
+    else
+    {
+        debug_output(opts->nlp_opts, "Trial iterate is NOT INSIDE of funnel\n", 1);
+    }
+    return accept_step;
+}
+
+static void ocp_nlp_sqp_compute_trial_iterate(ocp_nlp_config *config, ocp_nlp_dims *dims,
+            ocp_nlp_in *in, ocp_nlp_out *out, ocp_nlp_opts *opts, ocp_nlp_sqp_memory *sqp_mem,
+            ocp_nlp_workspace *work, double alpha)
+{
+    /* computes trial iterate in tmp_nlp_out */
+    int N = dims->N;
+    int *nv = dims->nv;
+    int *nx = dims->nx;
+    int *nu = dims->nu;
+    int *ni = dims->ni;
+    int *nz = dims->nz;
+
+    ocp_nlp_memory *mem = sqp_mem->nlp_mem;
+    ocp_nlp_out *tmp_nlp_out = work->tmp_nlp_out;
+
+    // Update primal, dual variables and slacks
+    int i = 0;
+    for (i = 0; i < N+1; i++)
+    {
+        // update primal variables
+        blasfeo_daxpy(nv[i], alpha, mem->qp_out->ux+i, 0, out->ux+i, 0, work->tmp_nlp_out->ux+i, 0);
```

Comment:
Not sure if this is correct.
In this line, `qp_out`, `out` and `tmp_nlp_out` are used.
I think it should be `tmp_nlp_out` instead of `out`.

Maybe don't introduce a new function and use `ocp_nlp_update_variables_sqp` with maybe minimal modifications?

Reply:
I think the update is correct. I merged the function with `ocp_nlp_update_variables_sqp`. 

One thing that puzzled me was that the default option for "full_step_dual"-update is always 0. Not sure, if this behavior is wanted. Normally SQP methods do a full step update of the duals even though the primals are updated by a line search. In any case, the funnel default option is changed to 1.